xicam/gui/bluesky/search.py

```diff
@@ -407,24 +408,80 @@ def __init__(self, *args, **kwargs):
         self.horizontalHeader().setSectionResizeMode(QHeaderView.ResizeToContents)
         self.setAlternatingRowColors(True)
 
+    def hide_hidden_columns(self):
+        hidden_columns = QSettings().value("catalog.columns.hidden")
+        if hidden_columns is None:
+            return
+        header = self.horizontalHeader()
+        current_column_names = [str(self.model().headerData(i, Qt.Horizontal)) for i in range(header.count())]
+        for column_name in hidden_columns:
+            if column_name in current_column_names:
+                header.setSectionHidden(current_column_names.index(column_name), True)
 
 class SearchWidget(QWidget):
     """
     Search input and results list
     """
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-
         self.catalog_selection_widget = CatalogSelectionWidget()
         self.search_input_widget = SearchInputWidget()
         self.search_results_widget = SearchResultsWidget()
-
         layout = QVBoxLayout()
         layout.addWidget(self.catalog_selection_widget)
         layout.addWidget(self.search_input_widget)
         layout.addWidget(self.search_results_widget)
         self.setLayout(layout)
 
+        header = self.search_results_widget.horizontalHeader()
+        header.setContextMenuPolicy(Qt.CustomContextMenu)
+        header.customContextMenuRequested.connect(self.header_menu)
+       
+    def hide_column(self, header, column_name, position):
+        '''
+        Hide a column, adding the column from the list of hidden columns in QSettings
+        '''
+        hidden_columns = QSettings().value("catalog.columns.hidden")
+        if hidden_columns is None:
+            hidden_columns = set()
+        hidden_columns.add(column_name)
+        QSettings().setValue("catalog.columns.hidden", hidden_columns)
+        header.setSectionHidden(position, True)
+
+    def unhide_column(self, header, column_name):
+        '''
+        Unhide a column, removing the column from the list of hidden columns in QSettings
+        '''
+        hidden_columns = QSettings().value("catalog.columns.hidden")
+        current_column_names = [str(header.model().headerData(i, Qt.Horizontal)) for i in range(header.count())]
+        position = current_column_names.index(column_name)
+        if hidden_columns is None:
+            hidden_columns = set()
+        else:
+            hidden_columns.remove(column_name)
```

Comment:
What do we want to have happen if the user tries to un-hide a column that isn't hidden? As written, that raises a `KeyError`. It's probably worth _considering_ doing nothing instead, which we could implement by using `discard` in place of `remove` here.

Reply:
I think that this state is ideally not possible; for now, we raise a new KeyError with a custom message. Thanks for your review!